zerver/tests/test_presence.py

```diff
@@ -139,6 +197,99 @@ def test_invalid_presence(self) -> None:
+        self.assertEqual(json["presence_last_update_id"], last_update_id)
+
+        # hamlet sends an active status. othello will next check presence and we'll
+        # want to verify he gets hamlet's update and nothing else.
+        self.login_user(hamlet)
+        params = dict(status="active", last_update_id=-1, slim_presence="true")
+        result = self.client_post("/json/users/me/presence", params)
+        json = self.assert_json_success(result)
+
+        # Make sure UserPresence.last_update_id is incremented.
+        self.assertEqual(
+            UserPresence.objects.latest("last_update_id").last_update_id, last_update_id + 1
+        )
+
+        # Now othello checks presence and should get hamlet's update.
+        self.login_user(othello)
+        params = dict(status="idle", slim_presence="true", last_update_id=last_update_id)
+        result = self.client_post("/json/users/me/presence", params)
+        json = self.assert_json_success(result)
+        self.assertEqual(set(json["presences"].keys()), {str(hamlet.id)})
+        self.assertEqual(json["presence_last_update_id"], last_update_id + 1)
+
+    def test_last_update_id_api_no_data_edge_cases(self) -> None:
+        hamlet = self.example_user("hamlet")
+
+        self.login_user(hamlet)
+
+        UserPresence.objects.all().delete()
+
+        params = dict(status="idle", last_update_id=-1, slim_presence="true")
+        # Make do_update_user_presence a noop. This simulates a production-like environment
+        # where the update is processed in a queue worker, so hamlet may not see his update
+        # reflected back to him in the response. Therefore it is as if there is no presence
+        # data.
+        # In such a situation, he should get his last_update_id=-1 back.
+        with mock.patch("zerver.worker.user_presence.do_update_user_presence"):
+            result = self.client_post("/json/users/me/presence", params)
+        json = self.assert_json_success(result)
+
+        self.assertEqual(set(json["presences"].keys()), set())
+        self.assertEqual(json["presence_last_update_id"], -1)
+        self.assertFalse(UserPresence.objects.exists())
+
+        # Now check the same, but hamlet doesn't pass last_update_id at all,
+        # like an old slim_presence client would due to an implementation
+        # prior to the introduction of last_update_id.
+        params = dict(status="idle", slim_presence="true")
+        with mock.patch("zerver.worker.user_presence.do_update_user_presence"):
+            result = self.client_post("/json/users/me/presence", params)
+        json = self.assert_json_success(result)
+        self.assertEqual(set(json["presences"].keys()), set())
+
+        # The server reflects the client's last_update_id back to it,
+        # so in this edge case it's a None.
+        self.assertEqual(json["presence_last_update_id"], None)
+        self.assertFalse(UserPresence.objects.exists())
```

Comment:
I don't love the idea of this field having a type that isn't `integer` from the standpoint of what clients need to validate. Probably worth discussing in the API design thread, but I think more rational might be to just fetch the current latest sequence number in that case? Not sure, but it'd be nice if clients don't need to think about parsing a parameter that might be `None`.

Reply:
A client that submitted `presence_last_update=None` (meaning they didn't send the param in the request at all; because e.g. this is the mobile app doing the fetch via `/register`) also doesn't even look at this value, so this might not matter? 

https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20response.20last_update_id.20when.20no.20new.20data

Reply:
Updated this to to be `self.assertEqual(json["presence_last_update_id"], -1)` as discussed in the thread